### ISSUE TITLE
Allow generic defaults for TSESLint.RuleModule?

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#4134]
- [x] ~I have tried restarting my IDE and the issue persists.~ N/A
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/TROUBLESHOOTING.md) and my problem is not listed.

**Repro**

Going off the code snippet here: https://github.com/typescript-eslint/typescript-eslint/pull/4124#discussion_r747998259

This gives an error [[playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQMoFFUBlgDt4C+cAZlBCHAOQACMAnmAKYDOAxlMGDALQsA2eGAHpGADyYcQjfAEM+3AK4xgfZpQDcAKE1ChcAOLTGHVnHpMqAJQV9GAWQgATG4wA8yOy2YyA5owCSjswANCgA8lzAELghKNa2OMwwRlAAfJRwUIwAjgrAWcxwAEaMMADujNJwAExwMriOcADMZgyMdVA+ClL4zAB0mmKQsHCs0UmZLgBcKBjYgn3x9k4ucAC8iJpwcLpwffuaBOpAA)]:

```ts
import { TSESLint } from '@typescript-eslint/experimental-utils';

// Generic type 'RuleModule<TMessageIds, TOptions, TRuleListener>' requires between 2 and 3 type arguments.
export const rule: TSESLint.RuleModule = {
```

Although I personally would use `RuleCreator`, or failing that specify those `TMessageIds` and `TOptions` type arguments, it is inconvenient for folks getting started with writing rules to have to specify them. It would make for smoother documentation if we didn't have to walk people through setting them up until they're needed.

Proposal:

```diff
interface RuleModule<
-  TMessageIds extends string,
+  TMessageIds extends string = string,
-  TOptions extends readonly unknown[],
+  TOptions extends readonly unknown[] = [],
```

**Versions**

| package                                 | version |
| --------------------------------------- | ------- |
| `@typescript-eslint/experimental-utils` | `5.3.1` |
| `TypeScript`                            | N/A |
| `node`                                  | N/A |

**Comment by bradzacher** [comment]
The problem with specifying defaults for `TMessageIds` is that then by default there is no typesafety over them.

```ts
const rule: TSESLint.RuleModule = {
  meta: {
    messages: {
      foo: 'FOOOO',
    },
  },
  create(context) {
    // no TS error :( :( :( 
    context.report({ messageId: 'obviouslyMissingId' });
  },
};

ruleTester.run('my-rule', rule, {
  valid: [],
  invalid: [
    {
      code: 'const x = 1;',
      // no TS error :( :( :( 
      errors: [{ messageId: 'obviouslyMissingId' }],
    },
  ],
});
```

Which is one of the major reasons we went with the rule creator function - because TS can infer the type for you!

```ts
const rule = createRule({
  meta: {
    messages: {
      foo: 'FOOOO',
      bar: 'BAAAR',
    },
  },
  create(context) {
    // TS error because it inferred TMessageIds as 'foo' | 'bar'
    context.report({ messageId: 'obviouslyMissingId' });
  },
});

ruleTester.run('my-rule', rule, {
  valid: [],
  invalid: [
    {
      code: 'const x = 1;',
      // TS error because it inferred TMessageIds as 'foo' | 'bar'
      errors: [{ messageId: 'obviouslyMissingId' }],
    },
  ],
});
```

---

Options, OTOH, is okay to specify a default for. If you forget to specify options then you just end up with the options `[]` and you can't do anything with the rule!
